language/diem-framework/modules/CRSN.move

```diff
@@ -0,0 +1,323 @@
+    public(friend) fun publish(account: &signer, min_nonce: u64, size: u64) {
+        assert(!has_crsn(Signer::address_of(account)), Errors::invalid_state(EHAS_CRSN));
+        assert(DiemVersion::get() > 3, Errors::invalid_state(ECRSN_NOT_SUPPORTED_AT_VERSION));
+        assert(size > 0, Errors::invalid_argument(EZERO_SIZE_CRSN));
+        move_to(account, CRSN {
+            min_nonce,
+            size,
+            slots: BitVector::new(size),
+        })
+    }
+
+    /// Record `sequence_nonce` under the `account`. Returns true if
+    /// `sequence_nonce` is accepted, returns false if the `sequence_nonce` is rejected.
+    public(friend) fun record(account: &signer, sequence_nonce: u64): bool
+    acquires CRSN {
+        let addr = Signer::address_of(account);
+        if (check(account, sequence_nonce)) {
+            // CRSN exists by `check`.
+            let crsn = borrow_global_mut<CRSN>(addr);
+            // accept nonce
+            let scaled_nonce = sequence_nonce - crsn.min_nonce;
+            BitVector::set(&mut crsn.slots, scaled_nonce);
+            shift_window_right(crsn);
+            return true
+        } else if (exists<CRSN>(addr)) { // window was force shifted in this transaction
+            let crsn = borrow_global<CRSN>(addr);
+            if (crsn.min_nonce > sequence_nonce) return true
+        };
+
+        false
+    }
+
+    /// A stateless version of `record`: returns `true` if the `sequence_nonce`
+    /// will be accepted, and `false` otherwise.
+    public(friend) fun check(account: &signer, sequence_nonce: u64): bool
+    acquires CRSN {
+        let addr = Signer::address_of(account);
+        assert(has_crsn(addr), Errors::invalid_state(ENO_CRSN));
+        let crsn = borrow_global_mut<CRSN>(addr);
+
+        // Don't accept if it's outside of the window
+        if ((sequence_nonce < crsn.min_nonce) ||
+            (sequence_nonce >= crsn.min_nonce + BitVector::length(&crsn.slots))) {
+            return false
+        };
+
+        // scaled nonce is the index in the window
+        let scaled_nonce = sequence_nonce - crsn.min_nonce;
+
+        // Bit already set, reject
+        if (BitVector::is_index_set(&crsn.slots, scaled_nonce)) return false;
+
+        // otherwise, accept
+        true
+
+    }
+    //  TODO/XXX: A force expiration will cause the transaction to abort due to the epilogue
+
+    /// Force expire transactions by forcibly shifting the window by
+    /// `shift_amount`. After the window has been shifted by `shift_amount` it is
+    /// then shifted over set bits as define by the `shift_window_right` function.
+    public fun force_expire(account: &signer, shift_amount: u64)
+    acquires CRSN {
+        if (shift_amount == 0) return;
+        let addr = Signer::address_of(account);
+        assert(has_crsn(addr), Errors::invalid_state(ENO_CRSN));
+        let crsn = borrow_global_mut<CRSN>(addr);
+
+        BitVector::shift_left(&mut crsn.slots, shift_amount);
```

Comment:
This aborts if `shift_amount` is bigger than the bitvector length. Are we ok with passing on that abort code, or do we want to do something like `assert(shift_amount > 0 && shift_amount < Vector::length(&crsn.slots), EINVALID_SHIFT_AMOUNT)`?

Reply:
I think if we are aborting with an `EINVALID_SHIFT_AMOUNT` for the shift-0 case, it would make sense to also use that same error code here and not pass the abort code fro mthe `shift_left` function. I'll update.

Reply:
As I'm updating this, I actually think we don't want to abort under the `shift_amount > #slots` as there could be cases where you might want to shift over more than a single window's worth at a time (and the `shift_left` operation in `BitVector` supports over-length shifts -- it just zeros out the whole bitvector).